RFCs/0014-TestCase-FullyQualifiedName.md

```diff
@@ -0,0 +1,48 @@
+# 0014 Guidelines for setting the FullyQualifiedName on TestCase
+
+## Summary
+This note clarifies the guidelines that test adapters should consider when setting the FullyQualifiedName on test cases before returning them from the discovery process.
+
+## Overview
+There are a broad variety of choices that current Test Adapter implementers have made when setting the FullyQualifiedName on TestCase objects created during test discovery. This note lays out some basic guidelines to make the usage of FullyQualifiedName more useful across adapters.
+
+## Usage of FullyQualifiedName
+The Test Platform uses FullyQualifiedName as a string moniker to name tests during discovery. Visual Studio may filter tests by FullyQualifiedName in order to more quickly filter and run specific tests. The guidelines below clarify the usage of FullyQualifiedName so that these optimized test execution scenarios can be realized.
+
+## Guidelines
+1. **FullyQualifiedName does NOT need to be unique across your test set.** Internally, TestCase objects are tracked by a unique ID (GUID) which can either be set by the Test Adapter implementor, or automatically generated otherwise. The FullyQualifiedName should be unique if possible, but it's not a requirement.
+1. **FullyQualifiedName should be deterministic** FullyQualifiedName should have the same value for the same test across different discovery sessions. This means that including random data, time-dependent data or environment specific data in the FullyQualifiedName should be avoided.
+1. **FullyQualifiedName is expected to be a reference to the method that implements the test** The Test Platform is expecting the FullyQualifiedName to refer to a particular test case. If there are multiple tests that are instantiated from a test method (data-driven tests for instance), then the FullyQualifiedName should distinguish these using parentheses to represent data parameters and angle brackets to represent type parameters. See below.
+
+## Recommended Form
+FullyQualifiedName is expected to be a reference to the TestCase method that implements the test (or tests). The FullyQualifiedName should be unique to a particular TestCase instance, if possible.
+
+The recommended forms of FullyQualifiedName are described by the following examples:
+
+    Namespace.Namespace.Class.Method
+    Namespace.Namespace.Class.Method()
+    Namespace.Namespace.Class<Int32>.Method()
+    Namespace.Namespace.Class<Int32>.Method<String>()
+    Namespace.Namespace.Class+InnerClass.Method
+    Namespace.Namespace.Class.Method(x: 1, msg: "string")
+    Namespace.Namespace.Class.Method(1, "string")
+    Namespace.Namespace.Class(init: "Value").Method(param: 23)
+    Namespace.Namespace.Class<String>("Value").Method(23)
+
+## Normalized Form
```

Comment:
I'm guessing this would be set by the platform. Would /testcasefilter be using the normalized form for filtering? Or would users be required to pass in the FullyQualifiedName format above for filtering?

Reply:
Currently this is handled by the TestWindow using the regex replacement logic. We did this so that there wouldn't be any requirement to change the Test Platform. There would be some benefits to doing this in the TestPlatform, but it's not an immediate requirement.

The main reason for this document is not to specify changes to the TestPlatform, but to document the normalization process for the benefit of test adapter developers.

Reply:
Sounds good Peter. One consequence of this however could be that all adapters start putting in test data in their FQN from the guidelines above and that could break filtering using `/testcasefilter:FullyQualifiedName="N.C.M(x: 24)"` when the method is adorned by a Random attribute where the data to the method is indeterminate by design.
To avoid such scenarios we might just want to recommend stripping off the data in FQN putting it in `DisplayName` instead. Thoughs?

Reply:
I think we will want to support test data in the FQN in the future, so that it will be possible to reference each individual test instance. In the case where non-deterministic data is included, I agree that it would not be possible to select that test by filter. I think I'd prefer to add NormalizedFQN as a filter property in that case to allow both scenarios.